src/lib/queue.ts

```diff
@@ -3,22 +3,32 @@ import { sleep, storageEngine } from '../helpers/utils';
 import VotesReport from './votesReport';
 
 export const queues = new Set<string>();
+const processingItems = new Set<string>();
```

Comment:
Looks good to me, however here are some possible issues:

- **Error handling:** Your `run` function's `catch` block simply logs the error and reruns the function. Enhance it to handle and possibly resolve errors before re-attempting.
  
- **Recursion:** Your `run` function uses recursion, which could risk a "Maximum call stack size exceeded" error due to lack of tail call optimization. Consider restructuring it.
  
- **Async/await usage in forEach loop:** The `forEach` loop in `run` doesn't await promises. If sequential execution is needed, use a `for...of` loop; if parallel execution is desired, use `Promise.all`.
  
- **Processing items:** In your `run` function, `processItem(item)` is not awaited. To ensure completion of processing, remember to `await` it or handle the returned promise.

Reply:
Example of improved run():
```js
async function run() {
  while (true) {
    try {
      log.info(`[queue] Poll queue (found ${queues.size} items)`);
      for (const item of queues) {
        if (processingItems.has(item)) {
          log.info(`[queue] Skip: ${item} is currently being processed`);
          continue;
        }
        await processItem(item);
      }
    } catch (e) {
      log.error(e);
    } finally {
      await sleep(15e3);
    }
  }
}

Reply:
- **Error Handling**: the function sole purpose is to prevent duplicate, like the `EE` function. In case of errors, task will just be discarded, like a regular call without the queue.
- **Recursion**: it's the same code used by the [hub](https://github.com/snapshot-labs/snapshot-hub/blob/d5ab9135b8fe5069fc7527d9ce7ab655cded0cdb/src/helpers/spaces.ts#L151) and [webhook](https://github.com/snapshot-labs/snapshot-webhook/blob/1d26c672067d6e16486595d3f633680756fc4f23/src/events.ts#L135). What do you mean by tail call optimization?
- **Async/await usage in forEach loop**: parallel execution is desired. What's the point of using Promise.all ? We will never wait for the resolution inside that forEach loop.
- **Processing items**: Awaiting for `processItem(item)` will make the loop sync. Without `await`, the task should still run, we just don't know the result, and we really don't care, the queue is just here to trigger task, and does not care for the returned value, unlike `EE`

Reply:
Makes sense. Thanks!